src/pytorch/pme/pmeCPU.cpp

```diff
@@ -0,0 +1,383 @@
+            int atom2 = pair[1][i];
+            bool include = (atom1 > -1);
+            for (int j = 0; include && j < maxExclusions && excl[atom1][j] >= atom2; j++)
+                if (excl[atom1][j] == atom2)
+                    include = false;
+            if (include) {
+                float invR = 1/r[i];
+                float alphaR = alpha*r[i];
+                float expAlphaRSqr = expf(-alphaR*alphaR);
+                float erfcAlphaR = erfcf(alphaR);
+                float prefactor = coulomb*invR;
+                float c1 = charge[atom1];
+                float c2 = charge[atom2];
+                energy += prefactor*erfcAlphaR*c1*c2;
+                chargeDeriv_a[atom1] += prefactor*erfcAlphaR*c2;
+                chargeDeriv_a[atom2] += prefactor*erfcAlphaR*c1;
+                float dEdR = prefactor*c1*c2*(erfcAlphaR+alphaR*expAlphaRSqr*M_2_SQRTPI)*invR*invR;
+                for (int j = 0; j < 3; j++) {
+                    posDeriv_a[atom1][j] -= dEdR*delta[i][j];
+                    posDeriv_a[atom2][j] += dEdR*delta[i][j];
+                }
+            }
+        }
+
+        // Subtract excluded interactions to compensate for the part that is
+        // incorrectly added in reciprocal space.
+
+        float dr[3];
+        for (int atom1 = 0; atom1 < numAtoms; atom1++) {
+            for (int i = 0; i < maxExclusions && excl[atom1][i] > atom1; i++) {
+                int atom2 = excl[atom1][i];
+                for (int j = 0; j < 3; j++)
+                    dr[j] = pos[atom1][j]-pos[atom2][j];
+                float rr = sqrt(dr[0]*dr[0] + dr[1]*dr[1] + dr[2]*dr[2]);
+                float invR = 1/rr;
+                float alphaR = alpha*rr;
+                float expAlphaRSqr = expf(-alphaR*alphaR);
+                float erfAlphaR = erff(alphaR);
+                float prefactor = coulomb*invR;
+                float c1 = charge[atom1];
+                float c2 = charge[atom2];
+                energy -= prefactor*erfAlphaR*c1*c2;
+                chargeDeriv_a[atom1] -= prefactor*erfAlphaR*c2;
+                chargeDeriv_a[atom2] -= prefactor*erfAlphaR*c1;
+                float dEdR = prefactor*c1*c2*(erfAlphaR-alphaR*expAlphaRSqr*M_2_SQRTPI)*invR*invR;
+                for (int j = 0; j < 3; j++) {
+                    posDeriv_a[atom1][j] += dEdR*dr[j];
+                    posDeriv_a[atom2][j] -= dEdR*dr[j];
+                }
+            }
+        }
+
+        // Store data for later use.
+
+        ctx->save_for_backward({posDeriv, chargeDeriv});
+        return {torch::tensor(energy, options)};
+    }
+
+    static tensor_list backward(AutogradContext *ctx, tensor_list grad_outputs) {
+        auto saved = ctx->get_saved_variables();
+        Tensor posDeriv = saved[0];
+        Tensor chargeDeriv = saved[1];
+        torch::Tensor ignore;
+        return {posDeriv*grad_outputs[0], chargeDeriv*grad_outputs[0], ignore, ignore, ignore, ignore, ignore, ignore};
+    }
+};
```

Comment:
I'm not sure it's possible to support second derivatives, except by writing code to compute them explicitly.  This corresponds to the last example in the page you linked.  Since the calculations are done by custom code, PyTorch doesn't know how to define gradients.

Reply:
This backwards is quite involved, the only way I see is to write the double backwards yourself, as you say.
The dangerous thing for me is that it will be silently wrong if you try to grad twice. Maybe there is a short way to at least raise an error if something tries to double backwards. Sadly, as far as I can see, even in the most simple form you need to implement a second autograd function like PME_Backwards.